Sources/GRPC/ConnectionBackoff.swift

```diff
@@ -0,0 +1,169 @@
+  public var maximumBackoff: TimeInterval
+
+  /// The backoff multiplier.
+  public var multiplier: Double
+
+  /// Backoff jitter; should be between 0 and 1.
+  public var jitter: Double
+
+  /// The minimum amount of time in seconds to try connecting.
+  public var minimumConnectionTimeout: TimeInterval
+
+  /// Creates a `ConnectionBackoff`.
+  ///
+  /// - Parameters:
+  ///   - initialBackoff: Initial backoff in seconds, defaults to 1.0.
+  ///   - maximumBackoff: Maximum backoff in seconds (prior to adding jitter), defaults to 120.0.
+  ///   - multiplier: Backoff multiplier, defaults to 1.6.
+  ///   - jitter: Backoff jitter, defaults to 0.2.
+  ///   - minimumConnectionTimeout: Minimum connection timeout in seconds, defaults to 20.0.
+  public init(
+    initialBackoff: TimeInterval = 1.0,
+    maximumBackoff: TimeInterval = 120.0,
+    multiplier: Double = 1.6,
+    jitter: Double = 0.2,
+    minimumConnectionTimeout: TimeInterval = 20.0
+  ) {
+    self.initialBackoff = initialBackoff
+    self.maximumBackoff = maximumBackoff
+    self.multiplier = multiplier
+    self.jitter = jitter
+    self.minimumConnectionTimeout = minimumConnectionTimeout
+  }
+
+  public func makeIterator() -> ConnectionBackoff.Iterator {
+    return Iterator(connectionBackoff: self)
+  }
+}
+
+/// An iterator for `ConnectionBackoff`.
+public class ConnectionBackoffIterator: IteratorProtocol {
+  public typealias Element = (timeout: TimeInterval, backoff: TimeInterval)
+
+  /// Creates a new connection backoff iterator with the given configuration.
+  public init(connectionBackoff: ConnectionBackoff) {
+    self.connectionBackoff = connectionBackoff
+    self.unjitteredBackoff = connectionBackoff.initialBackoff
+
+    // Generate the current and next elements. The first entry has no backoff (i.e. start the first
+    // attempt immediately). The second entry is not jittered.
+    //
+    // We generate `nextElement` before `currentElement` as we need to initialize `productionState`
+    // before can call `makeElement`.
+    if connectionBackoff.initialBackoff < connectionBackoff.maximumBackoff {
+      self.productionState = .producing
+      self.nextElement = self.makeElement(backoff: connectionBackoff.initialBackoff)
+    } else {
+      self.productionState = .stopped
+      self.nextElement = self.makeElement(backoff: connectionBackoff.maximumBackoff)
+    }
+
+    self.currentElement = self.makeElement(backoff: 0.0)
+  }
+
+  /// The configuration being used.
+  private let connectionBackoff: ConnectionBackoff
+
+  /// The backoff in seconds, without jitter.
+  private var unjitteredBackoff: TimeInterval
+
+  /// The state of production.
+  private var productionState: ProductionState
+
+  /// The current timeout and backoff waiting to be returned. This will become `nil` when we run
+  /// out of values to return.
+  private var currentElement: Element?
+
+  /// The next timeout and backoff to be returned.
+  private var nextElement: Element?
```

Comment:
Why do we need to store this? Couldn't we generate it on the fly when calling `next()`?

Reply:
Great question -- I was using this differently when starting the connection in a way which required the initial backoff to be zero (i.e. waiting and then trying instead of trying then waiting). Since the first _real_ backoff isn't jittered we needed to create it manually. Since that's not the case anymore we can simplify this.